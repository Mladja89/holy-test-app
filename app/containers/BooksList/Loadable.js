/**
 * Asynchronously loads the component for BooksList
 */

import loadable from '../../utils/loadable';

export default loadable(() => import('./index'));
