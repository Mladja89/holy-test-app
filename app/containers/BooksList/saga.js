import {
    take,
    call,
    put,
    fork,
} from 'redux-saga/effects';

import {
    SELECT_ROW,
} from './constants';
import {
    SET_SELECT_ROW,
} from '../../services/BE/books/constants';

export function* selectRow(payload) {
    yield put({ type: SET_SELECT_ROW, data: payload });
}

export function* selectRowFlow() {
    while (true) {
        const { payload } = yield take(SELECT_ROW);
        yield call(selectRow, payload);
    }
}

// The root saga is what we actually send to Redux's middleware. In here we fork
// each saga so that they are all 'active' and listening.
// Sagas are fired once at the start of an app and can be thought of as processes running
// in the background, watching actions dispatched to the store.
export default function* rootAppSaga() {
    yield fork(selectRowFlow);
}
