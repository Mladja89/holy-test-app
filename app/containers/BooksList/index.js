/*
 * BooksList
 *
 * This is the first thing users see of our App, at the '/' route
 *
 */

import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';

import MaterialTable from 'material-table';

import { selectRow } from './actions';

import Paper from '@material-ui/core/Paper';
import { displayDate } from '../../utils/formatters/datetime';
import { mapStateToProps } from '../../utils/mapStateToProps';
import { DAEMON } from '../../utils/constants';
import saga from './saga';
import injectSaga from '../../utils/injectSaga';

const BooksList = props => {
  const {
    books: { books, currentlySending: loadingBooks }
  } = props;

  if (!books.length || loadingBooks) {
    return 'Loading...';
  }

  const remappedBooks = books.map(book => {
    const BOOK = {
      ...book,
      writer: book.writer.name,
      dateOfBirth: displayDate(book.writer.dob),
    };
    const extendableBook = Object.assign({}, BOOK, { id: '' });

    return extendableBook;
  });

  var authorLookup = {}
  const tempObjUniqes = props.books.uniqueNames.map(author => author.writer.name)
  for (let key in tempObjUniqes) {
    authorLookup[tempObjUniqes[key]] = tempObjUniqes[key]
  }

  const [state, setState] = React.useState({
    columns: [
      {
        field: 'name',
        title: 'Title',
        filtering: false,
      },
      {
        field: 'writer',
        title: 'Author',
        lookup: {...authorLookup}
      },
      {
        field: 'yearOfPublishing',
        title: 'Year',
        type: 'numeric',
        filtering: false,
      },
      {
        field: 'numberOfPages',
        title: 'Number of pages',
        type: 'numeric',
        filtering: false,
      },
      {
        field: 'quantity',
        title: 'Quantity',
        type: 'numeric',
        filtering: false,
      },
    ],
    data: remappedBooks,
    tableRef: React.createRef(),
  });

  return (
    <Paper elevation={0}>
      <MaterialTable
        title="Books"
        columns={state.columns}
        data={state.data}
        tableRef={state.tableRef}
        onRowClick={(e, data) => props.selectRow(data)}
        options={{
          filtering: true
        }}
      />
    </Paper>
  );
}

const mapDispatchToProps = dispatch => ({
  selectRow: params => {
    dispatch(selectRow(params));
  }
});

export default compose(
  injectSaga({ key: 'BooksList', saga, mode: DAEMON }),
  connect(mapStateToProps, mapDispatchToProps),
)(BooksList);
