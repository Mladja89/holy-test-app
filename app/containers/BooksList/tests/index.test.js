import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';

import BooksList from '../index';

describe('<BooksList />', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale="en">
        <BooksList />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
