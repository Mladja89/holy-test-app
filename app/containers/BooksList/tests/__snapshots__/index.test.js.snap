// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BooksList /> should render and match the snapshot 1`] = `
<h1>
  <span>
    This is the BooksList container!
  </span>
</h1>
`;
